### PR TITLE
Add libgsm

### PR DESCRIPTION
**Pull request by wader**
Based on #155 by @mathieu-aubin

### CODE REVIEW

**Comment by mathieu-aubin** [comment]
This fails mainly because there is extra stuff in the library, stuff thats not needed. Also, `fPIE` shoudl be replaced with `fPIC` in my opinion...

i worked with my initial idea of the base64 encoded Makefile and this makes it compile just fine...

```
RUN   git clone "$LIBGSM_URL" &&  \
  cd libgsm && git checkout $LIBGSM_COMMIT && \
  echo 'Q0ZMQUdTCQk9CS1hbnNpIC1wZWRhbnRpYyAtYyAtTzMgLXMgLWZQSUMgLUROZWVkRnVuY3Rpb25Qcm90b3R5cGVzPTEgLVdhbGwgLVduby1jb21tZW50IC1EU0FTUiAtRFdBVjQ5IC1ETkRFQlVHIC1JLi9pbmMgLVdsLC16LHJlbHJvLC16LG5vdwpPQkpFQ1RTCQk9CXNyYy9hZGQubwkJXAoJCQlzcmMvY29kZS5vCQlcCgkJCXNyYy9kZWJ1Zy5vCQlcCgkJCXNyYy9kZWNvZGUubwkJXAoJCQlzcmMvbG9uZ190ZXJtLm8JCVwKCQkJc3JjL2xwYy5vCQlcCgkJCXNyYy9wcmVwcm9jZXNzLm8JXAoJCQlzcmMvcnBlLm8JCVwKCQkJc3JjL2dzbV9kZXN0cm95Lm8JXAoJCQlzcmMvZ3NtX2RlY29kZS5vCVwKCQkJc3JjL2dzbV9lbmNvZGUubwlcCgkJCXNyYy9nc21fZXhwbG9kZS5vCVwKCQkJc3JjL2dzbV9pbXBsb2RlLm8JXAoJCQlzcmMvZ3NtX2NyZWF0ZS5vCVwKCQkJc3JjL2dzbV9wcmludC5vCQlcCgkJCXNyYy9nc21fb3B0aW9uLm8JXAoJCQlzcmMvc2hvcnRfdGVybS5vCVwKCQkJc3JjL3RhYmxlLm8KCiUubzogc3JjLyUuYwoJCQlnY2MgJChDRkxBR1MpICQ/CgpsaWJnc20uYTogJChPQkpFQ1RTKQoJCQktcm0gbGliZ3NtLmEKCQkJYXIgY3IgbGliZ3NtLmEgJChPQkpFQ1RTKQoJCQlyYW5saWIgbGliZ3NtLmEKCmluc3RhbGw6IGxpYmdzbS5hCgkJCW1rZGlyIC1wIC91c3IvbG9jYWwvaW5jbHVkZS9nc20KCQkJY3AgaW5jL2dzbS5oIC91c3IvbG9jYWwvaW5jbHVkZS9nc20KCQkJY3AgaW5jL2dzbS5oIC91c3IvbG9jYWwvaW5jbHVkZQoJCQljcCBsaWJnc20uYSAvdXNyL2xvY2FsL2xpYgoKY2xlYW46CgkJCXJtIC1mIHNyYy8qLm8gKi5hCgpQSE9OWTogbGliZ3NtLmEK' | base64 -d > Makefile && \
  make -j$(nproc) install
```

which is this:

https://0x0.st/oPyx.txt

**Comment by wader** [comment]
> This fails mainly because there is extra stuff in the library, stuff thats not needed. Also, `fPIE` shoudl be replaced with `fPIC` in my opinion...

Think you'r correct, i even thought -fPIE was a link flag and that PIC was default now? but documentation seems to say they do similar things. Switch to -fPIC now.
> i worked with my initial idea of the base64 encoded Makefile and this makes it compile just fine...

The problem was `${SRC/.c/.o}` only substituting the first .c. I switch to just do *.o now.

**Comment by wader** [comment]
Souceforge having big problems today, maybe should try find other mirror for things

**Comment by mathieu-aubin** [comment]
> > This fails mainly because there is extra stuff in the library, stuff thats not needed. Also, `fPIE` shoudl be replaced with `fPIC` in my opinion...
> 
> Think you'r correct, i even thought -fPIE was a link flag and that PIC was default now? but documentation seems to say they do similar things. Switch to -fPIC now.
> 
> > i worked with my initial idea of the base64 encoded Makefile and this makes it compile just fine...
> 
> The problem was `${SRC/.c/.o}` only substituting the first .c. I switch to just do *.o now.

[Check this file](https://0x0.st/oPyx.txt), OBJECTS for libgsm do not comprise of all C files compiled.. there are extra ones for some other stuff not needed for the library.

The following works 

```Dockerfile
RUN \
  git clone "$LIBGSM_URL" && \
  cd libgsm && git checkout $LIBGSM_COMMIT && \
  # Makefile is garbage, hence use specific compile arguments and flags
  echo -e 'CFLAGS\t= -O3 -ansi -pedantic -c -s -static -fPIC -DNeedFunctionPrototypes=1 -Wall -Wno-comment -DSASR -DWAV49 -DNDEBUG -I./inc -Wl,-z,relro,-z,now\nOBJECTS\t= src/add.o src/code.o src/debug.o src/decode.o src/long_term.o src/lpc.o src/preprocess.o src/rpe.o src/gsm_destroy.o src/gsm_decode.o src/gsm_encode.o src/gsm_explode.o src/gsm_implode.o src/gsm_create.o src/gsm_print.o src/gsm_option.o src/short_term.o src/table.o\n\n%.o: src/%.c\n\tgcc $(CFLAGS) $?\n\nlibgsm.a: $(OBJECTS)\n\t-rm libgsm.a\n\tar cr libgsm.a $(OBJECTS)\n\tranlib libgsm.a\n\ninstall: libgsm.a\n\tmkdir -p /usr/local/include/gsm\n\tcp inc/gsm.h /usr/local/include/gsm\n\tcp inc/gsm.h /usr/local/include\n\tcp libgsm.a /usr/local/lib\n\nclean:\n\trm -f src/*.o *.a\n\nPHONY: libgsm.a' > Makefile && \
  make -j$(nproc) install
```
and FFmpeg can be linked with the resulting library

**Comment by mathieu-aubin** [comment]
maybe just remove `src/toast*.c` files before doing anything?

**Comment by wader** [comment]
> maybe just remove `src/toast*.c` files before doing anything? also, readme with gsm in TODO

Good idea, fixed.

File: Dockerfile [hunk 1]
```diff
@@ -50,8 +50,8 @@ RUN apk add --no-cache \
 # -static-libgcc is needed to make gcc not include gcc_s as "as-needed" shared library which
 # cmake will include as a implicit library.
 # other options to get hardened build (same as ffmpeg hardened)
-ARG CFLAGS="-O3 -static-libgcc -fno-strict-overflow -fstack-protector-all -fPIE"
-ARG CXXFLAGS="-O3 -static-libgcc -fno-strict-overflow -fstack-protector-all -fPIE"
+ARG CFLAGS="-O3 -static-libgcc -fno-strict-overflow -fstack-protector-all -fPIC"
```
Comment (mathieu-aubin) [inline Dockerfile:53]: shouldn't we add a `-s` (strip) here in order to strip shizz?

**Pull request merged**